test/integration/cdi/example-driver/kubeletplugin/draplugin.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubeletplugin
+
+import (
+	"fmt"
+
+	"google.golang.org/grpc"
+	"k8s.io/klog/v2"
+
+	drapbv1 "k8s.io/kubernetes/pkg/kubelet/apis/dra/v1alpha1"
+)
+
+type DRAPlugin interface {
+	// Stop ensures that all spawned goroutines are stopped and frees
+	// resources.
+	Stop()
+}
+
+// draPlugin combines the kubelet registration service and the DRA node plugin
+// service.
+type draPlugin struct {
+	registrar *nodeRegistrar
+	plugin    *grpcServer
+}
+
+// Start sets up two gRPC servers (one for registration, one for the DRA node
+// client).
+func Start(logger klog.Logger, driverName, endpoint, draAddress, pluginRegistrationPath string, nodeServer drapbv1.NodeServer) (result DRAPlugin, finalErr error) {
+	d := &draPlugin{}
+
+	// Run the node plugin gRPC server first to ensure that it is ready.
+	plugin, err := startGRPCServer(klog.LoggerWithName(logger, "dra"), endpoint, func(grpcServer *grpc.Server) {
+		drapbv1.RegisterNodeServer(grpcServer, nodeServer)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("start node client: %v", err)
+	}
+	d.plugin = plugin
+	defer func() {
+		// Clean up if we didn't finish succcessfully.
+		if finalErr != nil {
+			plugin.stop()
+		}
+	}()
+
+	// Now make it available to kubelet.
+	registrar, err := startRegistrar(klog.LoggerWithName(logger, "registrar"), driverName, draAddress, pluginRegistrationPath)
+	if err != nil {
+		return nil, fmt.Errorf("start registrar: %v", err)
+	}
+	d.registrar = registrar
```

Comment:
This is backwards from the recommended flow.

You are supposed to register first, which triggers your plugin directory to be created and an `fnotify` to be setup looking for socket files to be dropped into that directory. You then drop your socket file in that directory and start your gRPC server on it. Because of the `fnotify` that gets set up, the kubelet sees the socket file being dropped there and then knows to start listening on it.

Reply:
> This is backwards from the recommended flow.

Do you have a pointer to that recommendation?

> You are supposed to register first, which triggers your plugin directory to be created and an fnotify to be setup looking for socket files to be dropped into that directory. You then drop your socket file in that directory and start your gRPC server on it.

So the plugin is supposed to wait for a successful registration before creating its socket and not create the directory for that socket?

> Because of the fnotify that gets set up, the kubelet sees the socket file being dropped there and then knows to start listening on it.

kubelet needs to connect to that socket, not listen on it. It also works when the socket already exists before the fnotify. I don't know whether that is a feature of the fsnotify package or of kubelet, but unless this is specified in detail, the current implementation defines the plugin API and this functionality cannot be removed easily anymore.

I don't mind changing it, but the current approach seems simpler.

Reply:
> This is backwards from the recommended flow.

CSI drivers also don't follow the recommended flow: first the CSI driver creates the csi.sock, then node-driver-registrar handles the registration.

Reply:
This is the readme explaining the process: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pluginmanager/pluginwatcher/README.md

Though reading it again now, it's not clear if there is a defined order, and I agree that the way you are doing it now is more intuitive. So if it works, let's stick with it.

> kubelet needs to connect to that socket, not listen on it. 
Yes of course, sorry, brainfart ...